### PR TITLE
modified image preprocessing based on Chexpert Competition winner

### PR DESCRIPTION
**Pull request by zbuckley**
Chexpert Competition winner code: https://github.com/jfhealthcare/Chexpert

They had a mechanism for reducing the images, while maintaining the aspect ratios which I really liked. Also Using Histogram balancing, GaussianBlur, and RandomAffine(at least for the training set). 
Also changed our normalization and std error values to match.

### CODE REVIEW

File: Code/CheXpertDataset.py [hunk 1]
```diff
@@ -29,18 +30,13 @@
...
                  csv_file,
                  root_dir,
                  image_transform=None,
                  label_transform=None):
-        df = pd.read_csv(csv_file)
-
-        # Only use frontal images
-        self.df = df[df["Frontal/Lateral"] == "Frontal"]
-
+        self.df = pd.read_csv(csv_file)
```
Comment (zbuckley) [inline Code/CheXpertDataset.py:39]: Decided based on the Chexpert inputs we probably don't need to filter out the Frontal/Lateral view distinction after all. 

Believe this is because of the FixedRatioResize mechanism below.

File: Code/CheXpertDataset.py [hunk 2]
```diff
@@ -50,18 +46,24 @@ def __len__(self):
 
     def __getitem__(self, idx):
         image_path = self.df.iloc[idx, 0]
-        im = load_img(os.path.join(self.root_dir, image_path), target_size=(600, 600), interpolation='nearest')
+
+        # Updated based on: https://github.com/jfhealthcare/Chexpert/blob/47e1bf7f8ed2b95c8985ae43801ffffde92fa205/data/dataset.py#L104
+        # Image is grayscale here
+        image = cv2.imread(os.path.join(self.root_dir, image_path), 0)
+        image = Image.fromarray(image)
```
Comment (zbuckley) [inline Code/CheXpertDataset.py:53]: As comments suggest, did this to better match the competition winners preprocessing.

File: Code/CheXpertDataset.py [hunk 3]
```diff
@@ -76,11 +78,4 @@ def __call__(self, sample):
...
         sample = sample[label_columns].replace(-1, 1)
-        return sample
-
-class GreyScale_to_RGB:
-    def __call__(self, im):
-        return tf.image.grayscale_to_rgb(
-            im,
-            name=None
-        )
+        return sample
```
Comment (zbuckley) [inline Code/CheXpertDataset.py:81]:
```suggestion
        return sample

```

File: Code/CheXpertDataset.py [hunk 4]
```diff
@@ -77,10 +79,3 @@ class UOnesTransform:
     def __call__(self, sample):
         sample = sample[label_columns].replace(-1, 1)
         return sample
-
-class GreyScale_to_RGB:
-    def __call__(self, im):
-        return tf.image.grayscale_to_rgb(
-            im,
-            name=None
-        )
```
Comment (zbuckley) [inline Code/CheXpertDataset.py:86]: Weren't using this... ended up recreating it actually... but i'm using cv2 to match other preprocessing code.

File: Code/ImageTransforms.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            ratio = w * 1.0 / h
+            w_ = self.long_side
+            h_ = round(w_ / ratio)
+
+        image = cv2.resize(image, dsize=(w_, h_),
+                           interpolation=cv2.INTER_LINEAR)
+
+        image = self._border_pad(image)
+
+        return image
```
Comment (zbuckley) [inline Code/ImageTransforms.py:77]:
```suggestion
        return image

```

File: Code/training_common_utils.py [hunk 6]
```diff
@@ -4,15 +4,26 @@
...
     transforms.ToTensor(),
-    transforms.Normalize((0, 0, 0), (1.0, 1.0, 1.0))
+    transforms.Normalize((128, 128, 128), (64, 64, 64))
+])
+
+image_augmented_preprocessing = transforms.Compose([
+    transforms.RandomAffine(degrees=(-15, 15), translate=(0.05, 0.05),
+                            scale=(0.85, 1.05), fillcolor=128),
+    image_preprocessing
 ])
```
Comment (zbuckley) [inline Code/training_common_utils.py:27]: Order and settings for both the augmented and normal data preprocessing steps came from the order in which they were applied in the Chexpert winners solution.

File: Code/training_team8_uzeros.py [hunk 7]
```diff
@@ -40,7 +40,8 @@
 data_loader_params = {
     'batch_size': BATCH_SIZE,
     'shuffle': True,  # why not.. *shrug*?
-    'num_workers': 16}
+    'num_workers': 16,
+    'pin_memory': True}
```
Comment (zbuckley) [inline Code/training_team8_uzeros.py:44]: Pin memory setting as Kshitij suggested.

**Review by zbuckley — COMMENTED** [review]
LGTM. Let me know what you guys think.

File: Code/training_common_utils.py (hunk 6 above)
Comment (zbuckley) [reply Code/training_common_utils.py:27]: Note that image sizes should be 512x512 moving forward.

File: Code/CheXpertDataset.py (hunk 2 above)
Comment (josh-bernstein618) [reply Code/CheXpertDataset.py:53]: If you didn't do this, would you still need `PilToNumpy`?  Why not just directly convert `cv2` result to numpy?

File: Code/ImageTransforms.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+# assumes input data is a numpy array
+# should operate on Greyscale image
+class EqualizeHistogram:
+    def __call__(self, image):
+        return cv2.equalizeHist(image)
+
+
+# assumes input data is a numpy array
+# should operate on Grayscale image
+class Gray2RGB:
```
Comment (josh-bernstein618) [inline Code/ImageTransforms.py:20]: It's weird to me that this improves the results.  Given that the images are grayscale, I don't see why this would help, and I could see it hurting.  But if it actually helps .. I won't complain

File: Code/training_common_utils.py [hunk 9]
```diff
@@ -44,7 +55,7 @@ def build_data_loaders(base_label_transform, dlparams):
     data_training = ChexpertDataset(
         csv_file='../Data/CheXpert-v1.0-small/train.csv',
         root_dir='../Data',
-        image_transform=image_preprocessing,
+        image_transform=image_augmented_preprocessing,
```
Comment (josh-bernstein618) [inline Code/training_common_utils.py:58]: Wanna change line 51 as well?

File: Code/training_common_utils.py [hunk 10]
```diff
@@ -4,15 +4,26 @@
...
 from torchvision import transforms
 from torch.utils.data import Subset, DataLoader
 
 image_preprocessing = transforms.Compose([
-    # GreyScale_to_RGB(),
-    transforms.Resize((600, 600)),
+    PILToNumpy(),
+    EqualizeHistogram(),
+    Gray2RGB(),
+    FixRatioResize(512, 128),
```
Comment (josh-bernstein618) [inline Code/training_common_utils.py:17]: Does this need to be a tuple, like `((512, 128))`?  Also, it seems like line 61 expects a 3-tuple.  Maybe these aren't actually issues in Python, but wanted to bring up.

File: Code/CheXpertDataset.py (hunk 1 above)
Comment (josh-bernstein618) [reply Code/CheXpertDataset.py:39]: Do you know why this is?

**Review by josh-bernstein618 — COMMENTED** [review]
Looks mostly good, see a few questions/comments.

File: Code/CheXpertDataset.py (hunk 1 above)
Comment (zbuckley) [reply Code/CheXpertDataset.py:39]: As it'll handle resizing the images and keeping aspect ratio the model it scales it correctly. I didn't see anything in they're codebase that was removing the Lateral images. It looked like they simply trained on both. I do encourage you to take a second look though. I could certainly be wrong on that one.

Comment (zbuckley) [reply Code/CheXpertDataset.py:39]: you also may find this interesting... i hadn't though of the problem this way... they applied a different transform per label. 
https://github.com/jfhealthcare/Chexpert/blob/47e1bf7f8ed2b95c8985ae43801ffffde92fa205/data/dataset.py#L11

Took me awhile to figure out what it did. not a lot of comments in their code.

File: Code/CheXpertDataset.py (hunk 2 above)
Comment (zbuckley) [reply Code/CheXpertDataset.py:53]: That was actually on purpose... specifically to support the `RandomAffineTransform` I added for the augmented image preprocessing transform. Got that from their code, but I'm 85% sure using it requires the input to be a PIL Image object. 
I discovered some of that before.. it's why torchvision.transforms includes ToTensor and ToPIL... in some cases you have to switch back and forth a few times to get the job done.

File: Code/training_common_utils.py (hunk 10 above)
Comment (zbuckley) [reply Code/training_common_utils.py:17]: No? I wrote that one. It's in the ImageTransforms file now.

File: Code/training_common_utils.py (hunk 9 above)
Comment (zbuckley) [reply Code/training_common_utils.py:58]: No, but for a good reason. The data_training dataset is being initialized with image_augmented_preprocessing because the augmented preprocessing includes the `RandomAffineTransform` thing. The 'validation' set should be the original data, and shouldn't have anything randomized applied to it... instead sticking to the raw images (after being resized/normalized/histogram equalized, and and gaussian blurred. 

I need to read up more on the gaussian blur, but to my understanding these are all non-random transforms, and can thus be applied to all images before the model is ran. 

the randomAffine is a randomly applied transform causing (i think) rotation/translation/etc... along a couple different random distributions (i probably should double check the docs on that one too).

File: Code/training_common_utils.py (hunk 10 above)
Comment (zbuckley) [reply Code/training_common_utils.py:17]: Not sure which line 61 your referring too?
the normalize on line 20 is tuples but thats provided by the torchvision transforms, and isn't something we wrote.

Comment (josh-bernstein618) [reply Code/training_common_utils.py:17]: This line 61
https://github.com/ml2-final-project/Final-Project-Group8/pull/9/files#diff-37dd26a9587a0207f2418ef8c4462628R61

Also, I was concerned that this only takes one arg (aside from self)
https://github.com/ml2-final-project/Final-Project-Group8/pull/9/files#diff-37dd26a9587a0207f2418ef8c4462628R45

File: Code/ImageTransforms.py (hunk 8 above)
Comment (zbuckley) [reply Code/ImageTransforms.py:20]: Gray to RGB is required because the densenet121 architecture provided by torchvision is hardcoded to take 3 channels initially and adjusting/modifying that isn't trivial. 

Had a very good discussion with @kshitijbhat330 about it a few days ago. Thought you were on the call... but maybe not. I had looked into potentially changing the densenet model to take only a single channel at the first layer, but upon further digging it isn't something that the torchvision implementation made flexible. 
https://pytorch.org/docs/stable/_modules/torchvision/models/densenet.html#densenet121

File: Code/training_common_utils.py (hunk 10 above)
Comment (zbuckley) [reply Code/training_common_utils.py:17]: got it. remember that when we construct a composed transform we are constructing the objects (`__init__` function gets called), not actually invoking the transformation code (`__call__`). 
As such, I'm effectively using init to configure a thing that does the transformation, and only passing args to it when it needs some configuration regarding the behavior of the transform.
The only argument line 45 needs to take in is the image, as all our image transforms do. the `init` function on line 41 of the same file is the function called when we build the composed transform.
